src/Serializer/AbstractItemNormalizer.php

```diff
@@ -511,13 +513,22 @@ protected function getFactoryOptions(array $context): array
             $options['serializer_groups'] = (array) $context[self::GROUPS];
         }
 
-        if (isset($context['resource_class']) && $this->resourceClassResolver->isResourceClass($context['resource_class']) && $this->resourceMetadataCollectionFactory) {
-            $resourceClass = $this->resourceClassResolver->getResourceClass(null, $context['resource_class']); // fix for abstract classes and interfaces
-            // This is a hot spot, we should avoid calling this here but in many cases we can't
-            $operation = $context['operation'] ?? $this->resourceMetadataCollectionFactory->create($resourceClass)->getOperation($context['operation_name'] ?? null);
-            $options['normalization_groups'] = $operation->getNormalizationContext()['groups'] ?? null;
-            $options['denormalization_groups'] = $operation->getDenormalizationContext()['groups'] ?? null;
-            $options['operation_name'] = $operation->getName();
+        // This is a hot spot
+        if (isset($context['resource_class'])) {
+            $operationCacheKey = $context['resource_class'].($context['operation_name'] ?? '');
+            $operation = $context['operation'] ?? $this->localOperationCache[$operationCacheKey] ?? null;
+
+            if (!isset($this->localOperationCache[$operationCacheKey]) && !$operation && $this->resourceClassResolver->isResourceClass($context['resource_class']) && $this->resourceMetadataCollectionFactory) {
```

Comment:
Same here, `$this->resourceMetadataCollectionFactory()` is the slowest so should be called last.